prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +24,108 @@ def floatify(value, default=float("nan")):
+    # Group substations by ZIP code for a fallback for generators without coordinates
+    subs_zip_groupby = substations_with_xyz.groupby(["interconnect", "ZIP"])
+
+    # Create a KDTree for each combination of state and interconnect
+    state_trees = {
+        key: KDTree(np.array(substations_with_xyz.loc[sub_ids, "xyz"].tolist()))
+        for key, sub_ids in subs_state_lookup.items()
+    }
+    # Ensure that we have a tree for every generator
+    gens_state_groupby = generators_with_xyz.groupby(["interconnect", "state"])
+    missing_groups = set(gens_state_groupby.groups) - set(state_trees)
+    if len(missing_groups) > 0:
+        # There are some combinations of generator (interconnect, state) without subs
+        allowable_border_states = {"KS", "NE", "OK"}
+        for interconnect, state in missing_groups:
+            if state in allowable_border_states:
+                # Assume that the interconnect and state are correct
+                print(
+                    f"no substations within ({interconnect}, {state}), "
+                    f"will map generators to substations within {interconnect} instead"
+                )
+                # Find all substations for the interconnection
+                new_subs = substations_with_xyz.query("interconnect == @interconnect")
+                # Extend the 'true' combinations of (interconnect, state) with fakes
+                state_trees[(interconnect, state)] = KDTree(
+                    np.array(new_subs["xyz"].tolist())
+                )
+                subs_state_lookup[(interconnect, state)] = new_subs.index
+            else:
+                # Assume that the state is correct, the interconnect is wrong
+                print(
+                    f"no substations within ({interconnect}, {state}), "
+                    f"will map generators to substations within {state} instead"
+                )
+                (assumed,) = {
+                    interconnect
+                    for interconnect, state_list in const.interconnect2state.items()
+                    if interconnect not in {"ignore", "split"} and state in state_list
+                }
+                gens_to_fix = gens_state_groupby.get_group((interconnect, state)).index
+                generators_with_xyz.loc[gens_to_fix, "interconnect"] = assumed
+
+    # Query the appropriate tree for each generator to get the closest substation ID
+    sub_ids = generators_with_xyz.apply(
+        lambda x: get_sub_id_of_closest_substation(x, state_trees, subs_state_lookup),
         axis=1,
     )
-    return distance_to_subs.idxmin()
+    # For generators without coordinates, try to pick a substation with a matching ZIP
+    for g in generators.loc[sub_ids.isnull()].index:
+        try:
+            candidates = subs_zip_groupby.get_group(generators.loc[g, "ZIP"])
+            sub_ids.loc[g] = candidates.index[0]  # arbitrary choose the first one
```

Comment:
Just curious how many such cases? Our current logic flow is matching generator to substation first then to buses within the substation. If we match generator to buses directly, we can choose buses with proper `baseKV` here. Just thinking loud here maybe it's not doable.

Reply:
28 plants with no lat/lon in the 2019 Form 860, out of a total of 11,833.

We could definitely combine the logic of 'map generators to substation' and 'map generators to buses within substation' if we wanted to, since these are called one after another within `build_plant`.

Reply:
28/11833 is pretty minor. We can ignore it if they are not causing similar troubles that are described [here](https://app.zenhub.com/workspaces/renewable-energy-integration-5bfdd1184b5806bc2bf88976/issues/breakthrough-energy/renewableenergyproject/361) (I wasn't able to referencing the exact comment, but the one you uses Oconee nuclear plant as an example).